flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java

```diff
@@ -0,0 +1,301 @@
+
+	@Override
+	public void onAdded(List<KubernetesPod> pods) {
+		getMainThreadExecutor().execute(() -> {
+			for (KubernetesPod pod : pods) {
+				final String podName = pod.getName();
+				final CompletableFuture<KubernetesWorkerNode> requestResourceFuture = requestResourceFutures.remove(podName);
+
+				if (requestResourceFuture == null) {
+					log.debug("Ignore TaskManager pod that is already added: {}", podName);
+					continue;
+				}
+
+				log.info("Received new TaskManager pod: {}", podName);
+				requestResourceFuture.complete(new KubernetesWorkerNode(new ResourceID(podName)));
+			}
+		});
+	}
+
+	@Override
+	public void onModified(List<KubernetesPod> pods) {
+		terminatedPodsInMainThread(pods);
+	}
+
+	@Override
+	public void onDeleted(List<KubernetesPod> pods) {
+		terminatedPodsInMainThread(pods);
+	}
+
+	@Override
+	public void onError(List<KubernetesPod> pods) {
+		terminatedPodsInMainThread(pods);
+	}
+
+	@Override
+	public void handleFatalError(Throwable throwable) {
+		getMainThreadExecutor().execute(() -> getResourceEventHandler().onError(throwable));
+	}
+
+	// ------------------------------------------------------------------------
+	//  Internal
+	// ------------------------------------------------------------------------
+
+	private void recoverWorkerNodesFromPreviousAttempts() throws ResourceManagerException {
+		final List<KubernetesPod> podList = kubeClient.getPodsWithLabels(KubernetesUtils.getTaskManagerLabels(clusterId));
+		final List<KubernetesWorkerNode> recoveredWorkers = new ArrayList<>();
+
+		for (KubernetesPod pod : podList) {
+			final KubernetesWorkerNode worker = new KubernetesWorkerNode(new ResourceID(pod.getName()));
+			recoveredWorkers.add(worker);
+			final long attempt = worker.getAttempt();
+			if (attempt > currentMaxAttemptId) {
+				currentMaxAttemptId = attempt;
+			}
+		}
+
+		log.info("Recovered {} pods from previous attempts, current attempt id is {}.",
+				recoveredWorkers.size(),
+				++currentMaxAttemptId);
+
+		// Should not invoke resource event handler on the main thread executor.
+		// We are in the initializing thread. The main thread executor is not yet ready.
+		getResourceEventHandler().onPreviousAttemptWorkersRecovered(recoveredWorkers);
+	}
+
+	private KubernetesTaskManagerParameters createKubernetesTaskManagerParameters(TaskExecutorProcessSpec taskExecutorProcessSpec) {
+		final String podName = String.format(
+				TASK_MANAGER_POD_FORMAT,
+				clusterId,
+				currentMaxAttemptId,
+				++currentMaxPodId);
+
+		final ContaineredTaskManagerParameters taskManagerParameters =
+				ContaineredTaskManagerParameters.create(flinkConfig, taskExecutorProcessSpec);
+
+		final Configuration taskManagerConfig = new Configuration(flinkConfig);
+		taskManagerConfig.set(TaskManagerOptions.TASK_MANAGER_RESOURCE_ID, podName);
+
+		final String dynamicProperties =
+				BootstrapTools.getDynamicPropertiesAsString(flinkClientConfig, taskManagerConfig);
```

Comment:
Why is it necessary to reconstruct the `dynamicProperties` for `KubernetesTaskManagerParameters`? Wouldn't it be enough to pass on the `taskManagerConfig`?

Reply:
It compares `taskManagerConfig` with `flinkClientConfig` (the original version shipped from client), and send only the differences.